src/core/lib/iomgr/tcp_client_posix.cc

```diff
@@ -50,18 +51,174 @@
 
 extern grpc_core::TraceFlag grpc_tcp_trace;
 
-struct async_connect {
-  gpr_mu mu;
-  grpc_fd* fd;
-  grpc_timer alarm;
-  grpc_closure on_alarm;
-  int refs;
-  grpc_closure write_closure;
-  grpc_pollset_set* interested_parties;
-  std::string addr_str;
-  grpc_endpoint** ep;
-  grpc_closure* closure;
-  grpc_channel_args* channel_args;
+class AsyncConnect {
+ public:
+  AsyncConnect(grpc_fd* fd, grpc_pollset_set* interested_parties,
+               const grpc_resolved_address* addr, grpc_endpoint** ep,
+               grpc_closure* closure, const grpc_channel_args* channel_args,
+               const grpc_core::Timestamp& deadline)
+      : fd_(fd),
+        refs_(2),
+        interested_parties_(interested_parties),
+        addr_str_(grpc_sockaddr_to_uri(addr)),
+        ep_(ep),
+        closure_(closure),
+        channel_args_(grpc_channel_args_copy(channel_args)) {
+    GRPC_CLOSURE_INIT(&write_closure_, OnWritable, this,
+                      grpc_schedule_on_exec_ctx);
+    GRPC_CLOSURE_INIT(&on_alarm_, OnAlarm, this, grpc_schedule_on_exec_ctx);
+
+    if (GRPC_TRACE_FLAG_ENABLED(grpc_tcp_trace)) {
+      gpr_log(GPR_INFO, "CLIENT_CONNECT: %s: asynchronously connecting fd %p",
+              addr_str_.c_str(), fd_);
+    }
+
+    mu_.Lock();
+    grpc_timer_init(&alarm_, deadline, &on_alarm_);
+    grpc_fd_notify_on_write(fd_, &write_closure_);
+    mu_.Unlock();
+  }
+
+  ~AsyncConnect() { grpc_channel_args_destroy(channel_args_); }
+
+ private:
+  static void OnAlarm(void* acp, grpc_error_handle error) {
+    int done;
+    AsyncConnect* ac = static_cast<AsyncConnect*>(acp);
+    ac->mu_.Lock();
+    if (GRPC_TRACE_FLAG_ENABLED(grpc_tcp_trace)) {
+      gpr_log(GPR_INFO, "CLIENT_CONNECT: %s: on_alarm: error=%s",
+              ac->addr_str_.c_str(), grpc_error_std_string(error).c_str());
+    }
+    if (ac->fd_ != nullptr) {
+      grpc_fd_shutdown(
+          ac->fd_, GRPC_ERROR_CREATE_FROM_STATIC_STRING("connect() timed out"));
+    }
+    done = (--ac->refs_ == 0);
+    ac->mu_.Unlock();
+    if (done) {
+      delete ac;
+    }
+  }
+
+  static void OnWritable(void* acp, grpc_error_handle error) {
+    AsyncConnect* ac = static_cast<AsyncConnect*>(acp);
+    int so_error = 0;
+    socklen_t so_error_size;
+    int err;
+    int done;
+    grpc_endpoint** ep = ac->ep_;
+    grpc_closure* closure = ac->closure_;
+    grpc_fd* fd;
+    std::string addr_str = ac->addr_str_;
+
+    (void)GRPC_ERROR_REF(error);
+
+    ac->mu_.Lock();
```

Comment:
This is subjective, but for your consideration: I tend to prefer block-scope MutexLocks so that the start and end of the critical section are easier to see at a glance.